Add Scattering to `ParticleTracker`

JaydenR2305 opened the pull request:
Adds Coulomb scattering to the `ParticleTracker`

Merge #2712 and #2790 first

JaydenR2305 committed:
Remove `dev` from `index.rst`

JaydenR2305 committed:
Merge remote-tracking branch 'origin/main'

# Conflicts:
#	docs/whatsnew/index.rst

JaydenR2305 committed:
Rollover

JaydenR2305 committed:
Add changelog

JaydenR2305 committed:
Rollover

JaydenR2305 committed:
Rollover

JaydenR2305 committed:
Fix mypy directories

JaydenR2305 committed:
Fix mypy directories

JaydenR2305 committed:
Fix doc reference

JaydenR2305 committed:
Try `:noindex:`

JaydenR2305 committed:
Improve coverage

JaydenR2305 committed:
Improve coverage

JaydenR2305 committed:
Improve coverage

JaydenR2305 committed:
Fix documentation formatting
Add static stub files

JaydenR2305 committed:
Fix typo

JaydenR2305 committed:
Fix documentation

JaydenR2305 committed:
Fix documentation

JaydenR2305 committed:
Reorganize docs

JaydenR2305 committed:
Add new parameters to docstring

JaydenR2305 committed:
Update run method docstring

JaydenR2305 committed:
Update docs/api_static/plasmapy.simulation.particle_tracker.save_routines.rst

Co-authored-by: Nick Murphy <namurphy@cfa.harvard.edu>

JaydenR2305 committed:
Update docs/changelog/2024.5.0.rst

Co-authored-by: Nick Murphy <namurphy@cfa.harvard.edu>

JaydenR2305 committed:
fix docs?

JaydenR2305 committed:
Merge remote-tracking branch 'origin/src-reformat-particle-tracker-radiography' into src-reformat-particle-tracker-radiography

JaydenR2305 committed:
Basic setup

JaydenR2305 committed:
Keep track of energies
basic update routine

JaydenR2305 committed:
Merge remote-tracking branch 'refs/remotes/UPSTREAM/main' into particle-stopping

JaydenR2305 committed:
Reorganize stopping power-related routines in the atomic module

JaydenR2305 committed:
Finish `_push` implementation of basic stopping

JaydenR2305 committed:
Add basic changelog

JaydenR2305 committed:
Merge branch 'main' into particle-stopping

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

JaydenR2305 committed:
Clarify docstring

JaydenR2305 committed:
Improve comment narratives

JaydenR2305 committed:
Generalize interpolation of field components

JaydenR2305 committed:
Covnert `recognized_quantities` to a class method

JaydenR2305 committed:
Finish implementation of stopping

JaydenR2305 committed:
Add weird `H+` `p+` equivalence

JaydenR2305 committed:
Fix misnomer with `fraction_exited_threshold`

JaydenR2305 committed:
Add units to stopping interpolator

JaydenR2305 committed:
Update thresholds

JaydenR2305 committed:
Add `_stop_particles` logic at the end of the stopping event loop

JaydenR2305 committed:
Add `_stop_particles` logic at the end of the stopping event loop

JaydenR2305 committed:
Fix/add particle stopping logic at the end of the `_push` method
Correct speed mixup

JaydenR2305 committed:
Ignore stopped particles in final position distribution

JaydenR2305 committed:
Switch to `ParticleList` if statement for PSTAR

JaydenR2305 committed:
Add particle stopping tests

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

JaydenR2305 committed:
Add clarifying docstring to particle stopping test function

JaydenR2305 committed:
Add Bethe stopping
Add clarifying comments

JaydenR2305 committed:
fix typo

JaydenR2305 committed:
Merge branch 'main' into particle-stopping

JaydenR2305 committed:
Fix mypy list issue

JaydenR2305 committed:
Merge remote-tracking branch 'origin/particle-stopping' into particle-stopping

JaydenR2305 committed:
Remove `relativistic_beta_threshold` keyword

JaydenR2305 committed:
Fix docstring grammar

JaydenR2305 committed:
Fixed how `synthetic_radiograph` handles particles that stop before reaching the detector

JaydenR2305 committed:
Remove outdated relativity check

JaydenR2305 committed:
Simplify fsolve step in solving for proper time as a function of laboratory time
Remove unnecessary non-relativistic case from ExB trajectory fit

JaydenR2305 committed:
Consolidate `stopping_power` definition

JaydenR2305 committed:
More fsolve simplification

JaydenR2305 committed:
Add `keepdims=True` to norm functions

JaydenR2305 committed:
Remove Bethe stopping

JaydenR2305 committed:
Adaptive timestep

JaydenR2305 committed:
`_push` refactor

JaydenR2305 committed:
Docstring improvements

JaydenR2305 committed:
Merge branch 'main' into particle-stopping

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

JaydenR2305 committed:
Add square residual test to compare against classical trajectory

JaydenR2305 committed:
Instantiate integrator

JaydenR2305 committed:
Add test for mypy logic

JaydenR2305 committed:
Ternary operator for instantiating integrator

JaydenR2305 committed:
Add documentation for `return_interpolator`
Appease mypy

JaydenR2305 committed:
Add non-relativistic tests
Add check for `RelativityWarning` in respective regimes

JaydenR2305 committed:
Fix doc object references

JaydenR2305 committed:
Improve coverage

JaydenR2305 committed:
Improve coverage

JaydenR2305 committed:
Improve coverage

JaydenR2305 committed:
Improve coverage

JaydenR2305 committed:
remove debug

JaydenR2305 committed:
Improve changelog

JaydenR2305 committed:
Add `Bethe_stopping`

JaydenR2305 committed:
Add coverage for `Bethe_stopping`

JaydenR2305 committed:
Refactor `create_particles` to not set underlying variables managed by `load_particles`

JaydenR2305 committed:
"S" -> "dEdx"

JaydenR2305 committed:
Remove debug

JaydenR2305 committed:
Adjust signatures
Improve comments

JaydenR2305 committed:
Remove cognitive complexity comment

JaydenR2305 committed:
"instance" -> "subclass"

JaydenR2305 committed:
Improve coverage

JaydenR2305 committed:
Improve `stopping_power` docstring

JaydenR2305 committed:
Implement keyword argument for `add_stopping` to determine method

JaydenR2305 committed:
Update `fraction_exited_threshold` documentation and changelog entry

JaydenR2305 committed:
Improve coverage
Improve organization of `_interpolate_grid`
Update signature of `_update_position`, constructing E&B fields within the function body
Update return of `_interpolate_grid` to no longer include E&B fields

JaydenR2305 committed:
Implement Bethe stopping lite function to `ParticleTracker`

JaydenR2305 committed:
Reduce cognitive complexity of `add_stopping`
Implement new method for validating the parameters of `add_stopping`

JaydenR2305 committed:
Appease mypy

JaydenR2305 committed:
Appease mypy

JaydenR2305 committed:
Merge branch 'main' into particle-stopping

JaydenR2305 committed:
Add `PhysicsWarning` for low energy particles using Bethe stopping model
Improve coverage

JaydenR2305 committed:
Add breaking changelog

JaydenR2305 committed:
Decorate distribution functions in `Tracker` with `@staticmethod`

JaydenR2305 committed:
Fix Bethe stopping method reference

JaydenR2305 committed:
Merge branch 'main' into particle-stopping

JaydenR2305 committed:
Remove `max_theta` as a property in the `Tracker`

JaydenR2305 committed:
Improve performance of `test_NIST_particle_stopping`

JaydenR2305 committed:
Appease mypy

JaydenR2305 committed:
try `np.float64` -> `np.float32`

JaydenR2305 committed:
Revert to `np.ndarray` and add exception to mypy

JaydenR2305 committed:
Try float32 again

JaydenR2305 committed:
Try `np.flexible`

JaydenR2305 committed:
`np.flexible` -> `np.generic`

JaydenR2305 committed:
`np.flexible` -> `np.generic`

JaydenR2305 committed:
`np.generic` -> `np.number`

JaydenR2305 committed:
`np.number` -> `np.integer | np.floating`

JaydenR2305 committed:
`np.integer | np.floating` -> `np.integer[Any] | np.floating[Any]`

JaydenR2305 committed:
Add validate quantities

JaydenR2305 committed:
Merge branch 'main' into particle-stopping

JaydenR2305 committed:
Remove debug attributes

JaydenR2305 committed:
Merge branch 'main' into particle-stopping

JaydenR2305 committed:
First attempt at scattering

JaydenR2305 commented:
Current plan:
- Allow users to specify their own defined differential cross section for scattering
- Implement the Mott-Born formula and the exact Mott phase-shift formula as outlined in [this article](https://journals.aps.org/rmp/pdf/10.1103/RevModPhys.36.881)
- Specified cross sections will be numerically integrated to obtain mean-square rate of scattering:
![image](https://github.com/user-attachments/assets/f7058d83-acd5-4255-a5b5-54e983e15dcf)
where theta_min is-- in the above reference-- given to be twice the ratio of the impact parameter to the Debye length

pheuer commented:
Sounds like a good plan! Post some benchmarking of both models vs the ENDF data here for comparison when you get the models working!

JaydenR2305 committed:
Finish `add_scattering signature`

JaydenR2305 commented:
![image](https://github.com/user-attachments/assets/2079b8a6-9f9c-4ec9-9095-b83696536db7)
Here is what I have so far. This is for 5 MeV protons travelling through 2 cm of room temperature aluminum along with a Gaussian potential E field.

Are there any experimental data on differential scattering cross-sections for plamas that we can compare to?

pheuer commented:
> Here is what I have so far. This is for 5 MeV protons travelling through 2 cm of room temperature aluminum along with a Gaussian potential E field.

Good start! Presumably with stopping turned off, since that much Al would definitely stop the protons! I would do it without the E-field, since we can't simulate that in SRIM. 

> Are there any experimental data on differential scattering cross-sections for plamas that we can compare to?

Probably, but I'm not well versed on that literature - maybe you could poke around for some papers? If not I think we can benchmark in solids vs. SRIM to start - that's probably enough for the paper, although plasma scattering would be a great addition.

JaydenR2305 commented:
![image](https://github.com/user-attachments/assets/7d530ac9-1b3a-4117-9e05-432c54df2a68)
has the same form as
![image](https://github.com/user-attachments/assets/6f61262d-afaa-4547-86d0-5c3d89eed936)
with sigma = sqrt(<theta_0^2>)

JaydenR2305 commented:
Apparently I've been misunderstanding SRIM's definition of lateral straggling, which is given as:
![image](https://github.com/user-attachments/assets/e1c3db72-4c66-471c-b209-da0e58283cc4)
given that the ion beam is colinear with the x-axis

(see page 26 here for the definition of other statistical quantities relevant to scattering http://www.srim.org/SRIM/SRIM%2008.pdf) 
In general, the straggling is defined to be the second moment of the (cylindrically symmetric) distribution of ions about the center of the detector 🤷‍♂️

pheuer commented:
> With these adjustments I'm getting a value of 7.9 microns for the lateral straggling, which is at least _closer_ to the 8.65 micron expected value. One potential reason for this deviation is because I'm not currently removing back-scattered particles, which may represent a portion of the ion population.

Awesome, this looks great! Try excluding back-scattered particles to see if that helps, and test this over a number of different materials/densities. For the paper, I'm envisioning maybe a plot of lateral straggling vs. areal density of the target, and then plotting values for both SRIM and the tracker for 3-4 materials scanning across the areal density range? That should be a good validation. 

Also would be interesting to plot the analytical model there too, just to show where it diverges.

pheuer commented:
I'll start giving this code a rough review soon since it seems to be working - can you merge in main so the changes from the previous PR don't show up in the diff?

JaydenR2305 committed:
Finish user-defined scattering implementation

JaydenR2305 committed:
Merge branch 'main' into particle-scattering

JaydenR2305 committed:
Correct merge

JaydenR2305 commented:
> I'll start giving this code a rough review soon since it seems to be working - can you merge in main so the changes from the previous PR don't show up in the diff?

Should be good now!

pheuer commented on changelog/2712.feature.rst:
```diff
@@ -0,0 +1,3 @@
+Added the ability to enable particle stopping in the |ParticleTracker|.
+Added a ``return_interpolator`` keyword to `~plasmapy.particles.atomic.stopping_power` to allow the user to specify the return of an interpolator function (`~scipy.interpolate.CubicSpline` under the hood).
```
This line of the changelog was added in the last PR, right?

pheuer commented on src/plasmapy/simulation/particle_tracker/particle_tracker.py:
```diff
@@ -156,6 +159,7 @@ def __init__(
         dt_range=None,
         field_weighting="volume averaged",
         req_quantities=None,
+        seed: int | None = None,
```
good idea! Add this to the docstring

pheuer commented on src/plasmapy/simulation/particle_tracker/particle_tracker.py:
```diff
@@ -447,6 +456,8 @@ def load_particles(
         # Raise an error if the run method has already been called.
         self._enforce_order()
 
+        # TODO: it is an assumption that there exists only one specie of particles
```
Yes
```suggestion
```

pheuer commented on src/plasmapy/simulation/particle_tracker/particle_tracker.py:
```diff
@@ -447,6 +456,8 @@ def load_particles(
         self._enforce_order()
 
+        # TODO: it is an assumption that there exists only one specie of particles
+        #  is this a reasonable assumption?
```
```suggestion
```

pheuer commented on src/plasmapy/simulation/particle_tracker/particle_tracker.py:
```diff
@@ -584,6 +596,147 @@ def inner_Bethe_stopping(v, n_e):
 
+    # TODO: create a method to validate that these quantities have shape ngrids
+    # TODO: is this signature too much?
+    # TODO: find a way to create a type for the signature of the Callable parameters
```
Yes, this is the way to clean this up, not sure how it works though.

pheuer commented on src/plasmapy/simulation/particle_tracker/particle_tracker.py:
```diff
@@ -584,6 +596,147 @@ def inner_Bethe_stopping(v, n_e):
+            | None
+        ) = None,
+    ):
+        r"""
```
Expand docstring - lots of parameter details, physics stuff to explain here

pheuer commented on src/plasmapy/simulation/particle_tracker/particle_tracker.py:
```diff
@@ -584,6 +596,147 @@ def inner_Bethe_stopping(v, n_e):
+        self._target = target
+        self._do_scattering = True
+
+        # For now the integrals will be evaluated every time step
```
Is there an alternative to this, since the local density and temperature at each particle is different on each timestep?

JaydenR2305 replied:
Somewhere in the code I had a block commented out where I was essentially trying to construct a multidimensional interpolator that takes in v, T_s, and n_s and returns the mean-square rate of scattering. To do this I had to come up ranges for the parameters such that every possible combination of v, T_s, and n_s would be within the "domain" of the interpolator. The ranges (sampled in logspace) are v: 0 -> c , T_s: 0 -> max(T_s), and n_s: 0 -> max(n_s) where we cheat by taking the max value of temperature and number density encountered in the relevant grid. Does that sound like a reasonable approach?

pheuer replied:
I suppose the question is one of computation time, and consequently on how fine of a resolution for that grid is necessary for the initial computation, before the interpolation? If constructing the interpolator with sufficient resolution is fast enough, then this could work well.

pheuer commented on src/plasmapy/simulation/particle_tracker/particle_tracker.py:
```diff
@@ -675,17 +828,58 @@ def fract_entered(self):
+            for key, grid_value in self._total_grid_values.items()
+        }
+
+        # TODO: this seems awkward, maybe ``dt`` should be of shape ``nparticles``? as
```
Good idea - can you have the dt method cast it that way?

pheuer commented on src/plasmapy/simulation/particle_tracker/particle_tracker.py:
```diff
@@ -584,6 +596,147 @@ def inner_Bethe_stopping(v, n_e):
+    def add_scattering(
+        self,
+        target: ParticleLike,
+        differential_cross_section: (
```
I don't entirely understand where these callables come from, but here's an idea. 
I suggest: 

1) Make all of these parameters optional keywords to this function along with a 'method' keyword that chooses between the two. Make clear in the docstring which keywords are for which method, and which are for both. 

2) For the given method, ensure that the associated keywords are set, match the grid size, etc. 

3) Create the callable inside this function?

JaydenR2305 replied:
Just finished this, let me know what you think of how everything is documented :)

pheuer replied:
Two concerns still: 

1) The callables don't know what scattering type they are (?), so there's the potential for conflict. I still think it would be better, for the limited number of modes we have, to just have the callable selected internally based on a keyword that selects which mode to use? 

2) You should add all the other keywords as parameters, e.g. `T_s`, `n_s` to explain in detail what each one is. The docstring for the parameter can say `Only used for routines ....` for each parameter so it's clear which are required when. But we need to clearly define what they all are, and that's easier when they are all  listed as separate parameters in the normal format.

pheuer commented on src/plasmapy/simulation/particle_tracker/particle_tracker.py:
```diff
@@ -982,21 +1186,116 @@ def _update_velocity_stopping(self, summed_field_values):
+                        )
+                    ).si.value
+
+                # Calculate the minimum scattering angle based on screening
```
Write out all of what happens in both methods and how that is applied to the particles in latex math in the docstring for adding scattering. Bonus, you can then put that in the paper!

JaydenR2305 replied:
Sounds good. Did you have any recommended resources for scattering? I'm reading through Taylor's Classical Mechanics at the moment

pheuer replied:
I'll send you some stuff! I mostly meant just the equations showing how the mean scattering angle is obtained from the dsigma/domega, and provide references for the ENDF data vs. the analytic model (provide an equation for the analytic model). Basically just nicely write out what's in the code in more readable latex and explain where it comes from.

pheuer commented on src/plasmapy/simulation/particle_tracker/particle_tracker.py:
```diff
@@ -982,21 +1186,116 @@ def _update_velocity_stopping(self, summed_field_values):
+
+        # Begin by constructing the new velocities by adding the change in
+        # velocity to the current velocities. The direction of this vector will be
+        # correct but will need to be "rescaled" to unsure that the speed remains
```
```suggestion
        # correct but will need to be "rescaled" to ensure that the speed remains
```

pheuer reviewed (commented):
Looks pretty good so far! Figuring out how to create the callables in the user-accessible function that defines everything in a well documented easy to apply way is the most complex part I think. 

Make sure to describe the math fully in the docstring - as you've seen scattering is complex, so we need to be detailed about exactly what the code is doing.

JaydenR2305 committed:
Add `seed` to `ParticleTracker` docstring

JaydenR2305 committed:
Improve accuracy of error message
Remove unnecessary TODO
Improve commentary

JaydenR2305 committed:
Cast `dt` to be of length `nparticles` instead of `nparticles_tracked`

JaydenR2305 commented:
> Awesome, this looks great! Try excluding back-scattered particles to see if that helps

Just to make sure I'm understanding this correctly, back-scattered particles are where the angle between the velocity and src->detector vector is greater than 90 degrees?

pheuer commented:
> > Awesome, this looks great! Try excluding back-scattered particles to see if that helps
> 
> Just to make sure I'm understanding this correctly, back-scattered particles are where the angle between the velocity and src-to-detector vector is greater than 90 degrees?

Yes exactly

JaydenR2305 commented:
Just realizing that I forgot to move my stopping range test into the proper file before merging our other PR.. I'll add that alongside our lateral straggling tests here for scattering

JaydenR2305 commented:
![image](https://github.com/user-attachments/assets/073d6f85-de15-4251-82d2-1a0c8cca139e)
thank you IUPAC

JaydenR2305 committed:
Add Motte-Bethe equation to collision formulary
Improve documentation for `add_scattering`
Fix docstring for `load_particles`
Switch to keyword arguments for the signature of `scattering_routine`

JaydenR2305 committed:
Fix more `dt` stuff

JaydenR2305 committed:
Improve coverage

JaydenR2305 committed:
Fix push indexing

JaydenR2305 committed:
Implement support for non-proton beam particles

JaydenR2305 committed:
Improve coverage

JaydenR2305 committed:
Merge branch 'main' into particle-scattering

JaydenR2305 committed:
Fix NaN theta bug

JaydenR2305 committed:
Merge remote-tracking branch 'origin/particle-scattering' into particle-scattering

JaydenR2305 committed:
Allow hdf5 input to `synthetic_radiograph`
`str` <-> `Path` equivalence

JaydenR2305 committed:
remove debug

JaydenR2305 committed:
`output_file` -> `output_directory`

JaydenR2305 committed:
Get rid of warnings

JaydenR2305 committed:
Add warning for multi-grid scattering

JaydenR2305 committed:
`T_e` -> `T_i`
Improve error messages
Add parameter for NIST material string

JaydenR2305 commented:
@pheuer What are your thoughts on a `cast_iterable` keyword inside of the `validate_quantities` decorator? I'm running into the problem of implicitly assuming that the user isn't providing a scalar argument for things like `target_rho` inside of the `add_stopping` and `add_scattering` methods. The `cast_iterable` would convert any passed scalars into a (1,) NumPy array.

Also, should `validate_quantities` cast a single `ParticleLike` quantity into a `ParticleList` for arguments annotated with `ParticleList`? This is basically a similar case as above except with particle objects.

pheuer commented:
> @pheuer What are your thoughts on a `cast_iterable` keyword inside of the associated dictionary defined in the `validate_quantities` decorator? I'm running into the problem of implicitly assuming that the user isn't providing a scalar argument for things like `target_rho` inside of the `add_stopping` and `add_scattering` methods. The `cast_iterable` would convert any passed scalars into a (1,) NumPy array.

@namurphy 
I would say that would be scope creep for that decorator, since it's intended to validate u. Quantity objects. I wouldn't be opposed to creating a `cast_iteratable` decorator separately, although I might call it `cast_list` or `cast_tuple` to be explict about what the type is (or I guess there could be a keyword to `cast_iterable` to set that). 

> Also, should `validate_quantities` cast a single `ParticleLike` quantity into a `ParticleList` for arguments annotated with `ParticleList`? This is basically a similar case as above except with particle objects.

I feel like this could maybe be a keyword option for the @particle_input decorator? Again pinging @namurphy for their input.

JaydenR2305 committed:
Fix docstring
Update to relativistic kinetic energy

namurphy commented:
> @pheuer What are your thoughts on a `cast_iterable` keyword inside of the associated dictionary defined in the `validate_quantities` decorator? I'm running into the problem of implicitly assuming that the user isn't providing a scalar argument for things like `target_rho` inside of the `add_stopping` and `add_scattering` methods. The `cast_iterable` would convert any passed scalars into a (1,) NumPy array.



> Also, should `validate_quantities` cast a single `ParticleLike` quantity into a `ParticleList` for arguments annotated with `ParticleList`?

This would be under the purview of `@particle_input`




> > @pheuer What are your thoughts on a `cast_iterable` keyword inside of the associated dictionary defined in the `validate_quantities` decorator? I'm running into the problem of implicitly assuming that the user isn't providing a scalar argument for things like `target_rho` inside of the `add_stopping` and `add_scattering` methods. The `cast_iterable` would convert any passed scalars into a (1,) NumPy array.

> I wouldn't be opposed to creating a `cast_iteratable` decorator separately, although I might call it `cast_list` or `cast_tuple` to be explict about what the type is (or I guess there could be a keyword to `cast_iterable` to set that).

`validate_quantities` is currently quite complex, and I've found it difficult to add or change functionality in it (or even to add type hints). My inclination would be to create a separate but focused decorator that would include the needed functionality. 🤔 
 
> Also, should `validate_quantities` cast a single `ParticleLike` quantity into a `ParticleList` for arguments annotated with `ParticleList`?.

Yep, this would fall under the purview of `@particle_input`. One possibility would be that if the parameter is annotated with `ParticleListLike` instead of `ParticleLike`, then the corresponding argument would be transformed into a `ParticleList` instead of a `Particle | CustomParticle`.  I did a major refactoring of `@particle_input` a few years ago, so I'm hoping that it wouldn't be too difficult to modify this functionality! 😅

JaydenR2305 committed:
Convert electron density to a property intrinsic to the material grid
Fixed Bethe low energy warning
Converted to relativistic energy update equation

pheuer commented:
@JaydenR2305 If you choose to add these decorators, add them in a separate PR so we can review + merge them more quickly (rather than including them in this MR).

JaydenR2305 commented:
Another caveat is that an isotope needs to be specified for the target material in order to use the `mass_number` attribute-- so something like ``Particle("Al")`` wouldn't work but ``Particle("Al-27")`` would.

Maybe this could be a keyword argument to the ``particle_input`` decorator?

JaydenR2305 committed:
Make `Bethe_stopping_lite` return a positive number

JaydenR2305 committed:
Fix Bethe low speed warning

JaydenR2305 commented:
> Intuitively I feel like we should change it so `require` applies to all elements in a ParticleList type input?
Yep this is already the case apparently. The `particle_input` decorator passes `particlewise=True` to the `is_category` method of `ParticleList` which takes care of the rest

JaydenR2305 commented:
> `validate_quantities` is currently quite complex, and I've found it difficult to add or change functionality in it (or even to add type hints). My inclination would be to create a separate but focused decorator that would include the needed functionality. 🤔

The main problem with making another decorator is that the only parameters of interest (at least so far) are `Quantity` objects, meaning they wouldn't fall neatly into a category like `cast_list` or `cast_tuple`, since they are technically array values under the hood. Making `cast_iterator` a keyword argument of a decorator meant to deal with sanitizing `Quantity` objects made sense to me since we're technically just imposing another "sanitization" criteria be applied to the quantity before passing it to the decorated function.

JaydenR2305 committed:
Merge branch 'refs/heads/particle-list-input' into particle-scattering

JaydenR2305 commented:
@pheuer Also, what would you think about extending the `Particle` class to include array-like behavior by default? Meaning you would be able to instantiate a particle using a sequence of strings, then expressions like `2 * particle.atomic_number` would be supported without any changes to the code

JaydenR2305 commented:
also for the time being I've solved the array quantity problem by:
```py
if not isinstance(target_rho.value, np.ndarray):
    target_rho = np.asarray([target_rho.value]) * target_rho.unit
```

JaydenR2305 committed:
Docstring and parameter improvements

JaydenR2305 committed:
Merge branch 'main' into particle-scattering

pheuer commented:
> @pheuer Also, what would you think about extending the `Particle` class to include array-like behavior by default? 

This is an interesting idea - I think if you did this it would replace the ParticleList? Maybe this is a better way to do it? Worth opening an issue and tagging Nick to discuss, but for this PR I think the workaround you described is a good temporary solution.

JaydenR2305 committed:
Merge remote-tracking branch 'upstream/main' into particle-scattering

# Conflicts:
#	src/plasmapy/diagnostics/charged_particle_radiography/synthetic_radiography.py
#	src/plasmapy/simulation/particle_tracker/particle_tracker.py
#	src/plasmapy/simulation/particle_tracker/save_routines.py
#	tests/diagnostics/charged_particle_radiography/test_synthetic_radiography.py
#	tests/simulation/test_particle_tracker.py

pheuer commented on src/plasmapy/simulation/particle_tracker/particle_tracker.py:
```diff
@@ -546,6 +586,161 @@ def inner_Bethe_stopping(v, n_e):
+            A `~plasmapy.particles.particle_collections.ParticleListLike` representation of the target material(s).
+            The atomic mass of each particle must be specified (i.e. ``Al-27``).
+
+        scatter_routine : `callable`
```
Is there a reason to support both of these? It does make the code a lot more complicated and double the number of cases to test. Let's discuss.

pheuer commented on src/plasmapy/simulation/particle_tracker/particle_tracker.py:
```diff
@@ -1069,6 +1385,12 @@ def _push(self) -> None:
         if self._do_stopping:
             self._update_velocity_stopping()
 
+        if self.num_particles_tracked == 0:
```
When is this needed? Clarify in a comment

pheuer commented on tests/simulation/test_particle_tracker.py:
```diff
@@ -648,6 +653,171 @@ def test_particle_tracker_Bethe_warning(
         simulation.run()
 
 
+# def test_particle_tracker_scattering_warning(
```
Can we remove this block of commented out tests? Or is this waiting to be put back in?

pheuer commented on tests/simulation/test_particle_tracker.py:
```diff
@@ -648,6 +653,171 @@ def test_particle_tracker_Bethe_warning(
+    sim.add_scattering(
+        target_material,
+        target_density,
+        scatter_routine=Mott_Bethe_mean_squared_scattering,
```
This only tests the MSR option - what about the differential cross section option?

pheuer reviewed (commented):
Looking good! Let's discuss tomorrow

pheuer commented on src/plasmapy/simulation/particle_tracker/particle_tracker.py:
```diff
@@ -105,9 +110,24 @@ class to the ``run`` method.
 
         The default is 'volume averaged'.
 
+    req_quantities : `list` of `str`, default : `None`
```
And append rho to this list when yo uadd scattering

pheuer commented on src/plasmapy/simulation/particle_tracker/particle_tracker.py:
```diff
@@ -459,8 +492,15 @@ def _validate_stopping_inputs(
         return True
 
+    @particle_input(require="isotope")
+    @validate_quantities
```
"cold matter"

pheuer commented on src/plasmapy/simulation/particle_tracker/particle_tracker.py:
```diff
@@ -546,6 +586,161 @@ def inner_Bethe_stopping(v, n_e):
+    @validate_quantities
+    def add_scattering(
+        self,
+        target: ParticleListLike,
```
Need to re-think how target is specified: needs to work easily for something like a C4H8O2 or something

pheuer commented on tests/simulation/test_particle_tracker.py:
```diff
@@ -648,6 +653,171 @@ def test_particle_tracker_Bethe_warning(
+)
+@particle_input
+@validate_quantities
+def test_particle_tracker_ranges_straggling(
```
"stopping_and_scattering"

JaydenR2305 committed:
Fix docstring typos

JaydenR2305 committed:
Merge branch 'main' into particle-scattering

JaydenR2305 committed:
Fix `ParticleListLike` import

JaydenR2305 commented on src/plasmapy/formulary/collisions/misc.py:
```diff
@@ -267,7 +270,7 @@ def Bethe_stopping_lite(
     beta = v / _c.si.value
 
-    return np.asarray(
+    return -np.asarray(
```
Is this what you were looking for?
https://github.com/PlasmaPy/PlasmaPy/blob/d1ccae120252b88bb75f8c7d41bc418bc660edbe/src/plasmapy/simulation/particle_tracker/particle_tracker.py#L1013

Did you want me to add a comment about the minus sign to avoid confusion?

pheuer replied:
Great, probably a good idea to just mention that dEdx is by convention positive, so then this gets a minus sign

JaydenR2305 committed:
Improve `add_stopping` body

JaydenR2305 committed:
Delete old changelog

JaydenR2305 committed:
Clarifying comment on minus sign for `dE` calculation

JaydenR2305 committed:
stopping energy -> stopping power

JaydenR2305 committed:
Reorganize `ParticleTracker` constructor to reduce cognitive complexity
Remove unused code for computing classical scattering integral

JaydenR2305 committed:
Convert particle attributes to methods so they appear in documentation

JaydenR2305 committed:
Merge branch 'main' into particle-scattering

JaydenR2305 committed:
Add `num_particles` attribute

JaydenR2305 committed:
Clarify stopping interpolator comments in `add_stopping`

JaydenR2305 committed:
Update `test_particle_tracker` tests to match new function signature for `add_scattering`

JaydenR2305 committed:
Add changelog entry

JaydenR2305 committed:
Remove classical scattering integration

JaydenR2305 committed:
Improve `add_stopping` docstring

JaydenR2305 committed:
Remove invalid scattering routine meant for scattering in a plasma

JaydenR2305 committed:
Run pre-commit

JaydenR2305 committed:
Merge remote-tracking branch 'origin/particle-scattering' into particle-scattering

# Conflicts:
#	src/plasmapy/formulary/collisions/misc.py

JaydenR2305 committed:
Add rough sketch of new scattering method in formulary

JaydenR2305 committed:
Highland scattering sketch

JaydenR2305 committed:
Run pre-commit

JaydenR2305 committed:
Merge branch 'main' into particle-scattering

pheuer commented on src/plasmapy/formulary/collisions/misc.py:
```diff
@@ -436,6 +437,40 @@ def scattering_integrand(theta, B):
+    -------
+    theta_rms : `~astropy.units.Quantity`
+        The RMS scattering angle in radians.
+    """
```
Add a reference to the Highland paper in the docstring, and maybe a few notes on the regimes of relevance as described in the paper (e.g. E_s may not be constant below Z=20)

JaydenR2305 replied:
I just pushed a draft of the notes section for the `Highland_scattering` function in the collision formulary. I'm still not 100% on the details of the radiation length, but it seems like it's related to energy loss due to bremsstrahlung? I just wrote in the doc string that it was the relevant distance scale when we're talking about these "charge interactions"

pheuer replied:
I think it is used for protons/ions, note that the 'projectile mass' is refered to rather than just the electron mass

JaydenR2305 committed:
Fix typo in `epsilon` correction in Highland formula

JaydenR2305 committed:
Add notes section to `Highland_scattering` docstring

JaydenR2305 committed:
pray to the LaTeX god🙏

JaydenR2305 committed:
Use thickness instead of timestep
Add default value for atomic electron scattering

JaydenR2305 committed:
Finish implementing Bethe's review of Moliere multiple scattering

JaydenR2305 committed:
Merge remote-tracking branch 'origin/particle-scattering' into particle-scattering

# Conflicts:
#	src/plasmapy/formulary/collisions/misc.py

JaydenR2305 committed:
Finish implementing Bethe's review of Moliere multiple scattering

JaydenR2305 committed:
Implement ratio form of Eq. 22 in Bethe

pheuer commented on src/plasmapy/formulary/collisions/misc.py:
```diff
@@ -325,6 +334,157 @@ def Bethe_stopping(
+    return integral / factorial(n)
+
+
+def Bethe_Moliere_scattering(
```
I always just hear this called Moliere scattering- I think Bethe mostly just described Moliere's results in english?

pheuer commented on src/plasmapy/formulary/collisions/misc.py:
```diff
@@ -325,6 +328,80 @@ def Bethe_stopping(
+    average value for the distance of closest approach associated with the
+    classical Coulomb potential in the center of mass frame, given as :math:`b_0 = q_{\alpha}
+    q_{\beta} / 4 \pi \eps_0 \mu \bar{u}^2`, where :math:`\bar{u}` is the average relative velocity.
+    """
```
Also if this first PR just covers single elemental scattering, that would be a good start, we could add compounds next.

JaydenR2305 committed:
Merge remote-tracking branch 'origin/particle-scattering' into particle-scattering

JaydenR2305 committed:
`Bethe_Moliere_scattering` -> `Moliere_scattering`
Final rewrite of scattering routine

JaydenR2305 committed:
Add `return_rms` parameter to the `Moliere_scattering` function

JaydenR2305 committed:
Add interpolator to calculate integral

JaydenR2305 committed:
`quad` -> `quad_vec`

JaydenR2305 committed:
Test integration values provided by Bethe

JaydenR2305 committed:
Add atomic weight to recognized grid quantities

JaydenR2305 committed:
Improve docstring

pheuer commented on src/plasmapy/formulary/collisions/misc.py:
```diff
@@ -325,6 +464,213 @@ def Bethe_stopping(
+        return scattering_integrand
+    else:
+        # Eq. 30
+        theta_w = χ_c * np.sqrt(B)
```
Why not just use pi as the upper bound, and not rely on this approximation that neglects the higher order terms?

JaydenR2305 replied:
TODO: Figure out heuristic for including higher-order corrections when necessary

JaydenR2305 committed:
Add sketch of functions for stopping range and transmitted energy

JaydenR2305 committed:
Fix interpolator units

JaydenR2305 committed:
Finish thickness correction

JaydenR2305 committed:
Remove absolute tolerance for `f_mol` testing

JaydenR2305 committed:
Merge remote-tracking branch 'origin/particle-scattering' into particle-scattering

JaydenR2305 committed:
Simplify `Moliere_scattering` into a series of utility functions
Add support for compound targets

JaydenR2305 committed:
Add `speeds` attribute
Reduce cognitive complexity of `-update_velocity_stpoping`

JaydenR2305 committed:
Implement a tracked particle mask distinct from NaN velocity criterion

JaydenR2305 committed:
Consolidate NIST reads into a utility function

JaydenR2305 committed:
Implement consistent naming for NIST methods in `atomic`

JaydenR2305 committed:
Merge branch 'main' into particle-scattering

JaydenR2305 committed:
Fix `transmitted_energy`

JaydenR2305 committed:
Refactor to use new NIST methods
Calculate characteristic angles in the `_calculate_characteristic_angles` method
Add missing equation numbers

JaydenR2305 committed:
Appease linters

JaydenR2305 committed:
Appease linters

JaydenR2305 committed:
Appease linters

JaydenR2305 committed:
Improve documentation
Remove default energies return for NIST interpolators
Set `extrapolate=False` in `CubicSpline` constructor

JaydenR2305 committed:
Reconcile stopping logic by adding `_stopped_particle_mask_array`
Add new adaptive timestep based on stopping
Add `include_characteristic_angles` keyword argument to Moliere scattering method
Convert `range_type` to a required positional parameter
Add electron density to `_recognized_quantities_list`
Add `_H5FieldType`
Improve type checking in `particle_tracker`

JaydenR2305 commented on src/plasmapy/simulation/particle_tracker/particle_tracker.py:
```diff
@@ -844,6 +1056,25 @@ def _adaptive_dt(self) -> NDArray[np.float64] | float:
+            # max_delta_E = np.maximum(max_delta_E, (0.05 * u.MeV).to(u.J).value)
+
+            timestep = np.nan_to_num(
+                max_delta_E / (S[tracked_particles] * current_speeds), nan=np.inf
```
<img width="343" height="175" alt="Image" src="https://github.com/user-attachments/assets/ed04fb93-3984-4960-86b9-0eb76428c110" />

EDIT: dS/dx should read dS/dT, where T is the beam's kinetic energy

JaydenR2305 replied:
where S is the stoppingf power

JaydenR2305 replied:
computing a derivative from a spline in log-log space:
<img width="737" height="108" alt="image" src="https://github.com/user-attachments/assets/4236f4ae-7cba-4fd9-bdde-ed0bb38e0403" />
